### DISCUSSION TITLE
Error validation after onBlur. Clear error onChange until the next onBlur

### DISCUSSION THREAD

**Discussion by eakl** [react-hook-form/react-hook-form#8241] [Q&A] [answered]
Hello,

Is there a way to validate an input field after onBlur then clear the error message after refocus and the first key stroke then revalidate on the next onBlur.

I try to mimic the behavior of the Email field on the form located on this page https://monday.com/contact-us/

- Focus the input
- Type a wrong email format
- Blur
- Focus again, type any key
- Blur again

**Comment by Moshyfawn** [comment]
There's no build-in way for the described behavior. You can try setting
```js
useForm({
  mode: "onTouched"
})
```

From [RHF docs](https://react-hook-form.com/api/useform#props):
> Validation will trigger on the first blur event. After that, it will trigger on every change event.

[Here's a working example](https://codesandbox.io/s/modest-maxwell-c7qbs8?file=%2Fsrc%2FApp.tsx)

**Reply by eakl** [reply]
@Moshyfawn this is not the same behavior I'm looking for.
Your solution: Validation is triggered on the first blur event. After that, it is triggered on every change event.
What I'm looking for: Validation is triggered on the first blur event. On refocus AND **first keystroke**, error disappear until next blur event.

You can try the **Email** field on the contact form on this page https://monday.com/contact-us/

**Reply by Moshyfawn** [reply]
I get that `mode: "onTouched"` is not the desired behavior. RHF doesn't have a build-in way to handle your use-case.

Let me try something..

**Comment by Moshyfawn** [comment, marked as answer]
Here you go: you can
- Set the validation `mode: "onBlur"`
- Clear the field's error `onChange` via `clearErrors` form method if there's an error
```js
<input
  {...register('email', {
    onChange: () => {
      if (errors.email) {
        clearErrors('email')
      }
    },
    pattern: {
      value: emailRegex,
      message: 'Please, enter a valid email'
    }
  })}
/>
```

[Check out this example](https://codesandbox.io/s/intelligent-wescoff-emt39t?file=%2Fsrc%2FApp.tsx%3A576-837)

**Reply by eakl** [reply]
Thanks @Moshyfawn this is the expected behavior :)

**Reply by DmitryRevosafe** [reply]
Great answer @Moshyfawn, thank you. How this can be done in custom input component where I have no access to useForm params, only to useController

**Reply by Moshyfawn** [reply]
> Great answer @Moshyfawn, thank you. How this can be done in custom input component where I have no access to useForm params, only to useController

With controlled components, you control the even handlers and the value formatting. So, I'd say, something as simple as
```jsx
const ControlledInput = () => {
  const {
    field: { onChange },
    fieldState: { error }
  } = useController()
  
  return (
    <input
      onChange: newValue => {
        onChange(newValue)
        if (error) {
          // INFO: get the `clearErrors` method from either **props** or **useFormContext**
          clearErrors('fieldName')
        }
      }
    />
  )
}
```

_P.S. Haven't tested it in a live enviroment, just wrote the code in the notepad ;)_

**Reply by eakl** [reply]
I think it would be good to have a `clearError` (without the `s`) in useController for future versions

**Reply by Medsestrun** [reply]
Agree @eakl, it adds some extra work to pass additional prop or even wrap everything in FormProvider just to clear errors @Moshyfawn
